logic/concludable/ConjunctionConcludable.java

```diff
@@ -112,21 +119,89 @@ public Value asValue() {
-
-            // Find all roleplayer mapping combinations, which should have the form:
-            // Set<List<Pair<RelationConstraint.RolePlayer, RelationConstraint.RolePlayer>>> rolePlayerMappings
-            // For each, prune if there is no label intersection (or any other pruning, e.g. by value)
-            // Then build a Unification for each valid combination
-            return Stream.empty(); // TODO
+        public Stream<Map<Reference, Set<Reference>>> unify(HeadConcludable.Relation unifyWith) {
+            if (this.constraint().players().size() > unifyWith.constraint().players().size()) return Stream.empty();
+            Map<Reference, Set<Reference>> startingUnifier = new HashMap<>();
+            Optional<Map<Reference, Set<Reference>>> newUnifier;
+            if (constraint.owner().reference().isName()) {
+                newUnifier = updatedUnifier(this.constraint().owner(),
+                        unifyWith.constraint().owner(), startingUnifier);
+                if (newUnifier.isPresent()) startingUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            if (constraint.owner().isa().isPresent() && constraint.owner().isa().get().type().reference().isName()) {
+                assert unifyWith.constraint().owner().isa().isPresent();
+                newUnifier = updatedUnifier(this.constraint().owner().isa().get().type(),
+                        unifyWith.constraint().owner().isa().get().type(), startingUnifier);
+                if (newUnifier.isPresent()) startingUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            return matchRolesAndUnify(new ArrayList<>(this.constraint().players()), unifyWith.constraint().players(),
+                    new HashSet<>(), startingUnifier).stream();
+        }
+
+        private Set<Map<Reference, Set<Reference>>> matchRolesAndUnify(
+                List<RelationConstraint.RolePlayer> conjunctionRolePlayers,
+                List<RelationConstraint.RolePlayer> headRolePlayers,
+                Set<RelationConstraint.RolePlayer> visitedHeadRolePlayers,
+                Map<Reference, Set<Reference>> unifier
+        ) {
+            Set<Map<Reference, Set<Reference>>> allUnifiers = new HashSet<>();
+            if (conjunctionRolePlayers.isEmpty()) {
+                return set(unifier);
+            }
+            RelationConstraint.RolePlayer conj = conjunctionRolePlayers.remove(0);
+            for (RelationConstraint.RolePlayer head : headRolePlayers) {
+                if (visitedHeadRolePlayers.contains(head)) continue;
+                visitedHeadRolePlayers.add(head);
+                updatedUnifier(conj, head, unifier).ifPresent(newUnifier ->
+                        allUnifiers.addAll(matchRolesAndUnify(conjunctionRolePlayers,
+                                headRolePlayers, visitedHeadRolePlayers, newUnifier)));
```

Comment:
We recurse with the matched roles removed from the set of head and conjunction role-players, and keep going until we run out of conjunction players to match.